### DISCUSSION TITLE
'Multiple root elements detected' when script tag contains <div> element

### DISCUSSION THREAD

**Discussion by luttje** [livewire/livewire#7975] [7. Bugs (archived)] [answered]
### Livewire version

v3.4.4

### Laravel version

v10.44.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

When these conditions are met, a 'Multiple root elements detected' error occurs:
- Livewire root-level component is a Blade component
- Livewire component contains script tag  that contains an HTML string containing `<div></div>`

## Livewire component
**This code is what causes the error ⚠**
_resources/views/livewire/child-component.blade.php:_
```html
<x-page>
    <script>
        const test = '<div></div>';
    </script>
</x-page>
```

```php
<?php

// app/Livewire/ChildComponent.php

namespace App\Livewire;

use Livewire\Component;

class ChildComponent extends Component
{
    public function render()
    {
        return view('livewire.child-component');
    }
}
```

## Blade component used as the root element

```php
<?php

// app/View/Page.php

namespace App\View;

use Illuminate\Contracts\View\View;
use Illuminate\View\Component;

class Page extends Component
{
    public function render(): View
    {
        return view('page.index');
    }
}
```

### 
```html
<!-- resources/views/page/index.blade.php -->
<div>
    {{ $slot }}

    <div>test</div>
</div>
```

The `page` component is a Blade component, registered like this in AppServiceProvider:
```php
public function boot(): void
{
    Blade::components([
        Page::class => 'page',
    ]);
}
```

### resources/views/welcome.blade.php
```html
<html>
    <head>
        <title>My First Laravel App</title>
    </head>
    <body>
        @livewire('child-component')
    </body>
</html>
```

### Additional information

Full error: `Livewire only supports one HTML element per component. Multiple root elements detected for component: [child-component]`

While having a div element in the code does cause the problem, using a different element like h2 does not:
```html
<x-page>
    <script>
        const test = '<h2></h2>';
    </script>
</x-page>
```

I noticed this problem when using FilamentPHP, which brings users into this situation by generating this upon creating a page (which is a livewire component):
```html
<x-filament-panels::page>

</x-filament-panels::page>
```

If you add any HTML in a script tag you'll run into issues. I spent the last hour figuring out where to report this (myself/filamentphp/livewire/laravel) caused by following the instructions in [grapesjs](https://grapesjs.com/docs/getting-started.html#start-from-the-canvas), which have you write script tags containing HTML.

This problem does not occur when not using a blade component as the root components:
```html
<div>
    <x-filament-panels::page>
        <script>
            const test = '<div></div>';
        </script>
    </x-filament-panels::page>
</div>
```

#### Further experimentation

I believe the problem is somewhere inside `livewire/livewire/src/Features/SupportMultipleRootElementDetection/SupportMultipleRootElementDetection::getRootElementCount($html)`, because this code fails too:
```php
<?php 

$html = <<<HTML
<div>
	<script>
		const test = '<div></div>';
	</script>
	
	<div>
	</div>
</div>
HTML;

$dom = new \DOMDocument();
@$dom->loadHTML($html);

$body = $dom->getElementsByTagName('body')->item(0);
var_dump($body->nodeValue);

$count = 0;

foreach ($body->childNodes as $child) {
    if ($child->nodeType == XML_ELEMENT_NODE) {
        if ($child->tagName === 'script') continue;

        $count++;
    }
}

echo $count;
```
Outputs:
```
string(42) "
	
		const test = '<div>';
	
	
	
	
"
2
```

And this:
```php
var_dump($dom->saveHtml($body));
```
Outputs:
```
string(94) "<body><div>
	<script>
		const test = '<div></script></div>';
	
	
	<div>
	</div>
</body>"
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by luttje** [comment]
I guess this is where I find out that the problem is with `DOMDocument loadHTML`:
```php
<?php 

$html = <<<HTML
<div>
	<script>
		const test = '<div></div>';
	</script>
	
	<div>
	</div>
</div>
HTML;

$dom = new \DOMDocument();
var_dump($html);
@$dom->loadHTML($html);

var_dump($dom->saveHtml());
```

Outputs:
```
string(87) "<div>
	<script>
		const test = '<div></div>';
	</script>
	
	<div>
	</div>
</div>"
string(215) "<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div>
	<script>
		const test = '<div></script></div>';
	
	
	<div>
	</div>
</body></html>
"
```

However, reading the docs on that shows that livewire maybe should not use it:
![afbeelding](https://github.com/livewire/livewire/assets/2738114/539c2ad0-f113-4b79-b359-757cf950b4b6)
https://www.php.net/manual/en/domdocument.loadhtml.php

**Reply by luttje** [reply]
Any news on this? It feels clunky having to wrap my page components with another html element. Especially since it's only to fool the SupportMultipleRootElementDetection into giving a correct response (even though it parses the HTML all wrong)

**Comment by skeemer** [comment]
_resources/views/livewire/child-component.blade.php_ should be more like this:
```
   <div>
   </div>
    <script>
        const test = '<div></div>';
    </script>
```
The div tag is mandatory for a livewire component and you can't put anything beside it other than a script tag.

This also looks like you are trying to use a full page component and doing it wrong. https://livewire.laravel.com/docs/components#full-page-components

**Reply by luttje** [reply]
Thanks for your response! I think you're misunderstanding my problem. In my self-reply you can see that I narrowed the problem down to `DOMDocument loadHTML` using old parsing.

Whatever code I use in the child-component is irrelevant, since my minimum reproducible example shows the problem exists even without Livewire being there (but this code is what Livewire uses internally to detect multiple root components):
```php
<?php 

$html = <<<HTML
<div>
	<script>
		const test = '<div></div>';
	</script>
	
	<div>
	</div>
</div>
HTML;

$dom = new \DOMDocument();
var_dump($html);
@$dom->loadHTML($html);

var_dump($dom->saveHtml());
```

Outputs:
```
string(87) "<div>
	<script>
		const test = '<div></div>';
	</script>
	
	<div>
	</div>
</div>"
string(215) "<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div>
	<script>
		const test = '<div></script></div>';
	
	
	<div>
	</div>
</body></html>
"
```

Thanks for the effort though.

**Comment by skeemer** [comment, marked as answer]
It looks like we will get a fix in PHP 8.4: https://benjamincrozat.com/php-84#php-84s-dom-extension-now-supports-html5

But that's not dropping until November.

**Reply by luttje** [reply]
Oh that's sick! Good find, this is good enough for me (even if we have to wait a bit) so I'll mark your post as an answer.

Thanks for your help!
